### PR TITLE
[SMTChecker] Refactor CHC predicate instances

### PR DESCRIPTION
**Pull request by leonardoalt**
Depends on https://github.com/ethereum/solidity/pull/9763

Should split the commit into 3 commits:
- Error flag
- Predicate type
- Rest

### CODE REVIEW

**Comment by leonardoalt** [comment]
I've split the error flag part into the first commit.
In the end I didn't manage to split the `PredicateType` part because that causes changes in `setCurrentBlock` and `predicate` which need all the new stuff from `PredicateInstance`. Let me know if it's too hard to review like this.

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -725,20 +718,14 @@ void CHC::clearIndices(ContractDefinition const* _contract, FunctionDefinition c
 	}
 }
 
-void CHC::setCurrentBlock(
-	Predicate const& _block,
-	vector<smtutil::Expression> const* _arguments
-)
+void CHC::setCurrentBlock(Predicate const& _block)
```
Comment (leonardoalt) [inline libsolidity/formal/CHC.cpp:730]: @ekpyron this change is nice ;)

**Comment by leonardoalt** [comment]
This is ready for review.

File: libsolidity/formal/CHC.cpp [hunk 2]
```diff
@@ -215,16 +213,10 @@ void CHC::endVisit(FunctionDefinition const& _function)
 		{
 			string suffix = m_currentContract->name() + "_" + to_string(m_currentContract->id());
 			solAssert(m_currentContract, "");
-			auto constructorExit = createSymbolicBlock(
-				constructorSort(*m_currentContract),
-				"constructor_exit_" + suffix,
-				m_currentContract
-			);
-			connectBlocks(m_currentBlock, predicate(*constructorExit, currentFunctionVariables(*m_currentContract)));
+			auto constructorExit = createSymbolicBlock(constructorSort(*m_currentContract), "constructor_exit_" + suffix, PredicateType::ConstructorSummary, m_currentContract);
```
Comment (mijovic) [inline libsolidity/formal/CHC.cpp]:
```suggestion
			auto constructorExit = createSymbolicBlock(
				constructorSort(*m_currentContract),
				"constructor_exit_" + suffix,
				PredicateType::ConstructorSummary,
				m_currentContract
			);
```

File: libsolidity/formal/PredicateInstance.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	);
+}
+
+/// Currently it does not have arguments but it will have tx data in the future.
+smtutil::Expression implicitConstructor(Predicate const& _pred, ContractDefinition const&, EncodingContext&)
+{
+	return _pred({});
+}
+
+smtutil::Expression function(Predicate const& _pred, FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context)
```
Comment (mijovic) [inline libsolidity/formal/PredicateInstance.cpp]:
```suggestion
smtutil::Expression function(
	Predicate const& _pred,
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
)
```

File: libsolidity/formal/PredicateInstance.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+{
+	return _pred(currentFunctionVariables(_function, _contract, _context));
+}
+
+smtutil::Expression functionBlock(Predicate const& _pred, FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context)
```
Comment (mijovic) [inline libsolidity/formal/PredicateInstance.cpp]:
```suggestion
smtutil::Expression functionBlock(
	Predicate const& _pred,
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
)
```

File: libsolidity/formal/PredicateInstance.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+vector<smtutil::Expression> currentStateVariables(ContractDefinition const& _contract, EncodingContext& _context)
+{
+	return applyMap(
+		SMTEncoder::stateVariablesIncludingInheritedAndPrivate(_contract),
+		[&](auto _var) { return _context.variable(*_var)->currentValue(); }
+	);
+}
+
+vector<smtutil::Expression> currentFunctionVariables(FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context)
```
Comment (mijovic) [inline libsolidity/formal/PredicateInstance.cpp]:
```suggestion
vector<smtutil::Expression> currentFunctionVariables(
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
)
```

File: libsolidity/formal/PredicateInstance.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * The predicates follow the specification in PredicateSort.h.
+ * */
+
+smtutil::Expression interface(Predicate const& _pred, ContractDefinition const& _contract, EncodingContext& _context);
+
+smtutil::Expression constructor(Predicate const& _pred, ContractDefinition const& _contract, EncodingContext& _context);
+
+smtutil::Expression implicitConstructor(Predicate const& _pred, ContractDefinition const& _contract, EncodingContext& _context);
+
+smtutil::Expression function(Predicate const& _pred, FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context);
```
Comment (mijovic) [inline libsolidity/formal/PredicateInstance.h]:
```suggestion
smtutil::Expression function(
	Predicate const& _pred,
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
);
```

File: libsolidity/formal/PredicateInstance.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+smtutil::Expression functionBlock(Predicate const& _pred, FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context);
```
Comment (mijovic) [inline libsolidity/formal/PredicateInstance.h]:
```suggestion
smtutil::Expression functionBlock(
	Predicate const& _pred,
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
);
```

File: libsolidity/formal/PredicateInstance.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+std::vector<smtutil::Expression> initialStateVariables(ContractDefinition const& _contract, EncodingContext& _context);
+
+std::vector<smtutil::Expression> stateVariablesAtIndex(unsigned _index, ContractDefinition const& _contract, EncodingContext& _context);
+
+std::vector<smtutil::Expression> currentStateVariables(ContractDefinition const& _contract, EncodingContext& _context);
+
+std::vector<smtutil::Expression> currentFunctionVariables(FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context);
+
+std::vector<smtutil::Expression> currentBlockVariables(FunctionDefinition const& _function, ContractDefinition const* _contract, EncodingContext& _context);
```
Comment (mijovic) [inline libsolidity/formal/PredicateInstance.h]:
```suggestion
std::vector<smtutil::Expression> currentFunctionVariables(
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
);

std::vector<smtutil::Expression> currentBlockVariables(
	FunctionDefinition const& _function,
	ContractDefinition const* _contract,
	EncodingContext& _context
);
```

**Review by mijovic — COMMENTED** [review]
It looks good. Only few problems with line length...

File: test/libsolidity/smtCheckerTests/loops/while_loop_array_assignment_storage_storage.sol [hunk 9]
```diff
@@ -23,4 +23,3 @@ contract LoopFor2 {
 // ----
 // Warning 6328: (320-339): CHC: Assertion violation happens here.
 // Warning 6328: (343-362): CHC: Assertion violation happens here.
-// Warning 4661: (296-316): BMC: Assertion violation happens here.
```
Comment (mijovic) [inline test/libsolidity/smtCheckerTests/loops/while_loop_array_assignment_storage_storage.sol:26]: What made it not asserting anymore?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/loops/while_loop_array_assignment_storage_storage.sol:26]: The first assertion is actually safe, so ideally this is the expected result. Sometimes if the encoding changes ever so slightly (even if just predicate/variable names) the result might also change, in which case CHC wasn't (before) able to solve this, and BMC reports it as a false positive.

**Comment by leonardoalt** [comment]
Updated (adding the changes to the second commit)

File: test/libsolidity/smtCheckerTests/loops/while_loop_array_assignment_storage_storage.sol (hunk 9 above)
Comment (mijovic) [reply test/libsolidity/smtCheckerTests/loops/while_loop_array_assignment_storage_storage.sol:26]: Maybe than deleting comment (about false positive) just before first assertion?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/loops/while_loop_array_assignment_storage_storage.sol:26]: Ok, will delete.

**Comment by leonardoalt** [comment]
@mijovic updated

**Review by mijovic — DISMISSED** [review]

**Comment by leonardoalt** [comment]
@mijovic I had forgotten to update the test expectation, can you reapprove?

**Review by mijovic — APPROVED** [review]

**Pull request merged**